### PR TITLE
Revive subset of nupic.core for nupic.research

### PR DESCRIPTION
**Pull request by mrcslws**
See individual commit messages. This change gives us Python 3 access to a `SparseMatrixConnections` which can be used for networks with dynamically added dendritic segments. With this change, all of our other C++ data structures / algorithms are low-hanging fruit.

Remaining work that is out of scope for this PR:

- Add python bindings for every class and method that we intend to use.
  - Related task: get the `bindings/py/tests` to pass. I only fixed `connections_test.py`.
- Set up the CI, update CircleCI config, etc.

### CODE REVIEW

**Comment by lscheinkman** [comment]
In the current setup, `nupic` was converted from `package` to `namespace`. Do you need `bindings/py/nupic` to be a package?
Can we delete this file?
`bindings/py/nupic/__init__.py`

**Comment by mrcslws** [comment]
Good call, that fixed a confusing issue I was seeing. (`pip install -e .` caused `import nupic.bindings` to fail, while `pip install .` worked). I don't fully understand why the `__import__('pkg_resources').declare_namespace(__name__)` inside `__init__.py` wasn't still making it a namespace. Maybe a Python 2 -> 3 difference.

File: README.md [hunk 1]
```diff
@@ -1,130 +1,62 @@
...
 
-    # While still in $NUPIC_CORE/build/scripts
+    # While still in nupic.research.core/build/scripts
     make install
 
 #### Run the tests:
 
-    cd $NUPIC_CORE/build/release/bin
-    ./cpp_region_test
+    cd ../bin
```
Comment (lscheinkman) [inline README.md]: The `cmake` command above will install the binaries under the `../release` folder.
It should be `cd ../release/bin`

**Comment by lscheinkman** [comment]
After running `./unit_tests` the  tests passed but the stderr output had the following errors:
```
ERR:  CHECK FAILED: "numDesired > 0" Not enough columns (10) for desired density (0.02). [nupic.research.core/src/nupic/algorithms/SpatialPooler.cpp line 974]
ERR:  Number of column dimensions must be greater than 0 [nupic.research.core/src/nupic/algorithms/TemporalMemory.cpp line 105]
ERR:  Number of cells per column must be greater than 0 [nupic.research.core/src/nupic/algorithms/TemporalMemory.cpp line 109]
ERR:  CHECK FAILED: "isSortedWithoutDuplicates(activeColumns, activeColumns + activeColumnsSize)" The activeColumns must be a sorted list of indices without duplicates. [nupic.research.core/src/nupic/algorithms/TemporalMemory.cpp line 456]
ERR:  CHECK FAILED: "isSortedWithoutDuplicates(activeColumns, activeColumns + activeColumnsSize)" The activeColumns must be a sorted list of indices without duplicates. [nupic.research.core/src/nupic/algorithms/TemporalMemory.cpp line 456]
ERR:  Invalid cell 16384 [nupic.research.core/src/nupic/algorithms/TemporalMemory.cpp line 635]
ERR:  Invalid cell 4294967295 [nupic.research.core/src/nupic/algorithms/TemporalMemory.cpp line 635]
ERR:  input (9.9) less than range [10, 20] [nupic.research.core/src/nupic/encoders/ScalarEncoder.cpp line 81]
ERR:  input (20.1) greater than range [10, 20] [nupic.research.core/src/nupic/encoders/ScalarEncoder.cpp line 88]
ERR:  input 9.9 not within range [10, 20) [nupic.research.core/src/nupic/encoders/ScalarEncoder.cpp line 148]
ERR:  input 20 not within range [10, 20) [nupic.research.core/src/nupic/encoders/ScalarEncoder.cpp line 148]
ERR:  Invalid NTA_BasicType -1 used in array constructor [nupic.research.core/src/nupic/ntypes/ArrayBase.cpp line 59]
ERR:  allocateBuffer -- buffer already set. Use releaseBuffer first [nupic.research.core/src/nupic/ntypes/ArrayBase.cpp line 75]
ERR:  allocateBuffer -- buffer already set. Use releaseBuffer first [nupic.research.core/src/nupic/ntypes/ArrayBase.cpp line 75]
ERR:  setBuffer -- buffer already set. Use releaseBuffer first [nupic.research.core/src/nupic/ntypes/ArrayBase.cpp line 91]
ERR:  CHECK FAILED: "index < vec_.size()"  [nupic.research.core/src/nupic/ntypes/Collection.cpp line 65]
ERR:  No item named: blah [nupic.research.core/src/nupic/ntypes/Collection.cpp line 86]
ERR:  CHECK FAILED: "index < vec_.size()"  [nupic.research.core/src/nupic/ntypes/Collection.cpp line 65]
ERR:  No item named: blah [nupic.research.core/src/nupic/ntypes/Collection.cpp line 86]
ERR:  CHECK FAILED: "index < vec_.size()"  [nupic.research.core/src/nupic/ntypes/Collection.cpp line 65]
ERR:  No item named: blah [nupic.research.core/src/nupic/ntypes/Collection.cpp line 86]
ERR:  CHECK FAILED: "index < vec_.size()"  [nupic.research.core/src/nupic/ntypes/Collection.cpp line 65]
ERR:  No item named: blah [nupic.research.core/src/nupic/ntypes/Collection.cpp line 86]
ERR:  Unable to add item '0' to collection because it already exists [nupic.research.core/src/nupic/ntypes/Collection.cpp line 95]
ERR:  Unable to add item '1' to collection because it already exists [nupic.research.core/src/nupic/ntypes/Collection.cpp line 95]
ERR:  Unable to add item '2' to collection because it already exists [nupic.research.core/src/nupic/ntypes/Collection.cpp line 95]
ERR:  No item named '4' in collection [nupic.research.core/src/nupic/ntypes/Collection.cpp line 111]
ERR:  No item named '1' in collection [nupic.research.core/src/nupic/ntypes/Collection.cpp line 111]
ERR:  Unable to add item '0' to collection because it already exists [nupic.research.core/src/nupic/ntypes/Collection.cpp line 95]
ERR:  Unable to add item '1' to collection because it already exists [nupic.research.core/src/nupic/ntypes/Collection.cpp line 95]
ERR:  Unable to add item '2' to collection because it already exists [nupic.research.core/src/nupic/ntypes/Collection.cpp line 95]
ERR:  No item named '1' in collection [nupic.research.core/src/nupic/ntypes/Collection.cpp line 111]
ERR:  No item named '0' in collection [nupic.research.core/src/nupic/ntypes/Collection.cpp line 111]
ERR:  Attempt to get count from dimensions [unspecified] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 50]
ERR:  Bad request for dimension 0 on [unspecified] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 63]
ERR:  Invalid coordinate [1 2] for Dimensions [unspecified] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 128]
ERR:  Attempt to get count from dimensions [unspecified] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 50]
ERR:  Bad request for dimension 0 on [unspecified] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 63]
ERR:  Invalid coordinate index 0 of 0 is too large for region dimensions [dontcare] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 142]
ERR:  Attempt to get count from dimensions [dontcare] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 50]
ERR:  Invalid coordinate index 0 of 1 is too large for region dimensions [1 0] (invalid) [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 142]
ERR:  Attempt to get count from invalid dimensions [1 0] (invalid) [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 55]
ERR:  Bad request for dimension 2 on [1 0] (invalid) [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 63]
ERR:  Bad request for dimension 2 on [2 3] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 63]
ERR:  Bad request for dimension 2 on [2 5] [nupic.research.core/src/nupic/ntypes/Dimensions.cpp line 63]
ERR:  CHECK FAILED: "cP < endP_" MemParser::get(string&) - EOF [nupic.research.core/src/nupic/ntypes/MemParser.cpp line 175]
ERR:  CHECK FAILED: "cP < endP_" MemParser::get(string&) - EOF [nupic.research.core/src/nupic/ntypes/MemParser.cpp line 175]
ERR:  CHECK FAILED: "cP < endP_" MemParser::get(string&) - EOF [nupic.research.core/src/nupic/ntypes/MemParser.cpp line 175]
ERR:  CHECK FAILED: "endP != startP_ && endP <= endP_" MemParser::get(long&) - parse error, not a valid integer [nupic.research.core/src/nupic/ntypes/MemParser.cpp line 129]
ERR:  Attempt to enable node with index 5 which is larger than the number of nodes 4 [nupic.research.core/src/nupic/ntypes/NodeSet.hpp line 70]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 71]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 71]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt32"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 63]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_Int32"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 67]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_Int64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 75]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt32"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 63]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 71]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 71]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 71]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 71]
ERR:  CHECK FAILED: "theType_ == NTA_BasicType_UInt64"  [nupic.research.core/src/nupic/ntypes/Scalar.cpp line 71]
ERR:  CHECK FAILED: "category_ == arrayCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 74]
ERR:  CHECK FAILED: "category_ == stringCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 79]
ERR:  Attempt to access scalar of type Int32 as type UInt32 [nupic.research.core/src/nupic/ntypes/Value.cpp line 86]
ERR:  CHECK FAILED: "category_ == scalarCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 69]
ERR:  CHECK FAILED: "category_ == stringCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 79]
ERR:  CHECK FAILED: "category_ == scalarCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 84]
ERR:  CHECK FAILED: "category_ == scalarCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 69]
ERR:  CHECK FAILED: "category_ == arrayCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 74]
ERR:  CHECK FAILED: "category_ == scalarCategory"  [nupic.research.core/src/nupic/ntypes/Value.cpp line 84]
ERR:  Key 'scalar' specified twice [nupic.research.core/src/nupic/ntypes/Value.cpp line 112]
ERR:  population size must be greater than number of choices [nupic.research.core/src/nupic/utils/Random.hpp line 126]

```

**Comment by lscheinkman** [comment]
When installing the bindings i got the following deprecation error:
```
  ERROR: Failed building wheel for pycapnp
  Running setup.py clean for pycapnp
Failed to build pycapnp
Installing collected packages: coverage, pytest-cov, pycapnp, pybind11
    Running setup.py install for pycapnp ... done
  DEPRECATION: pycapnp was installed using the legacy 'setup.py install' method, because a wheel could not be built for it. A possible replacement is to fix the wheel build issue reported above. You can find discussion regarding this at https://github.com/pypa/pip/issues/8368.
Successfully installed coverage-6.2 pybind11-2.9.0 pycapnp-0.6.3 pytest-cov-3.0.0
```

**Comment by mrcslws** [comment]
That's the expected `unit_tests` output, I think. Look at any nupic.core CI results, e.g. this one: https://github.com/numenta/nupic.core/commit/4de16752bbe4c80675789e619af8433c0f18e5d6

It's the same. Confusing logging, for sure.

File: README.md (hunk 1 above)
Comment (mrcslws) [reply README.md]: Nice catch.

**Comment by mrcslws** [comment]
Thanks for walking through the install steps, I reproduce the pycapnp install failure. It works when I use a more recent version. For this PR, I'm just setting it to use a recent version, and in a future change I'll really look at getting the capnp bindings working.

**Review by lscheinkman — APPROVED** [review]

**Pull request merged**